crypto/pkcs7/pkcs7_asn1.c

```diff
@@ -0,0 +1,186 @@
+        ADB_ENTRY(NID_pkcs7_enveloped, ASN1_EXP_OPT(PKCS7, d.enveloped, PKCS7_ENVELOPE, 0)),
+        ADB_ENTRY(NID_pkcs7_signedAndEnveloped, ASN1_EXP_OPT(PKCS7, d.signed_and_enveloped, PKCS7_SIGN_ENVELOPE, 0)),
+        ADB_ENTRY(NID_pkcs7_digest, ASN1_EXP_OPT(PKCS7, d.digest, PKCS7_DIGEST, 0)),
+        ADB_ENTRY(NID_pkcs7_encrypted, ASN1_EXP_OPT(PKCS7, d.encrypted, PKCS7_ENCRYPT, 0))
+} ASN1_ADB_END(PKCS7, 0, type, 0, &p7default_tt, NULL);
+
+ASN1_SEQUENCE(PKCS7) = {
+    ASN1_SIMPLE(PKCS7, type, ASN1_OBJECT),
+    ASN1_ADB_OBJECT(PKCS7)
+} ASN1_SEQUENCE_END(PKCS7)
+
+IMPLEMENT_ASN1_ALLOC_FUNCTIONS(PKCS7)
+
+PKCS7 *d2i_PKCS7(PKCS7 **a, const unsigned char **in, long len) {
+  uint8_t *der_bytes = NULL;
+  PKCS7 *ret = NULL;
+  CBS cbs, cbs_der;
+
+  if (!in) {
+    return NULL;
+  }
+
+  CBS_init(&cbs, *in, len);
+  // |CBS_asn1_ber_to_der| will allocate memory and point |der_bytes| to it.
+  // we're responsible for freeing this below.
+  if (!CBS_asn1_ber_to_der(&cbs, &cbs_der, &der_bytes)) {
+      goto err;
+  }
+
+  // |CBS_asn1_ber_to_der| will set |der_bytes| to NULL if it doesn't detect
+  // any convertible BER elements in |in|.
+  if (der_bytes == NULL) {
+    ret = (PKCS7 *) ASN1_item_d2i(
+        (ASN1_VALUE **) a, in, len,
+        ASN1_ITEM_rptr(PKCS7)
+    );
+  } else {
+    // |ASN1_item_d2i| will increment the input pointer by |der_len| length, so
+    // save off another pointer so we can free |der_bytes| at the end of this
+    // function.
+    uint8_t *der_bytes_ptr = der_bytes;
+    size_t der_len = CBS_len(&cbs_der);
+    ret = (PKCS7 *) ASN1_item_d2i(
+        (ASN1_VALUE **) a, (const uint8_t**) &der_bytes_ptr,
+        der_len, ASN1_ITEM_rptr(PKCS7)
+    );
+    // Advance |*in| by however many bytes |ASN1_item_d2i| advanced
+    // |der_bytes_ptr|
+    *in += der_bytes_ptr - der_bytes;
+  }
+
+err:
+  OPENSSL_free(der_bytes);
+  der_bytes = NULL;
+  return ret;
+}
+
+int i2d_PKCS7(PKCS7 *a, unsigned char **out) {
+    return ASN1_item_i2d((ASN1_VALUE *)a, out, ASN1_ITEM_rptr(PKCS7));
+}
+
+IMPLEMENT_ASN1_DUP_FUNCTION(PKCS7)
+
+ASN1_SEQUENCE(PKCS7_SIGNED) = {
+        ASN1_SIMPLE(PKCS7_SIGNED, version, ASN1_INTEGER),
+        ASN1_SET_OF(PKCS7_SIGNED, md_algs, X509_ALGOR),
+        ASN1_SIMPLE(PKCS7_SIGNED, contents, PKCS7),
+        ASN1_IMP_SEQUENCE_OF_OPT(PKCS7_SIGNED, cert, X509, 0),
+        ASN1_IMP_SET_OF_OPT(PKCS7_SIGNED, crl, X509_CRL, 1),
+        ASN1_SET_OF(PKCS7_SIGNED, signer_info, PKCS7_SIGNER_INFO)
+} ASN1_SEQUENCE_END(PKCS7_SIGNED)
+
+IMPLEMENT_ASN1_FUNCTIONS(PKCS7_SIGNED)
+
+ASN1_SEQUENCE(PKCS7_ISSUER_AND_SERIAL) = {
+        ASN1_SIMPLE(PKCS7_ISSUER_AND_SERIAL, issuer, X509_NAME),
+        ASN1_SIMPLE(PKCS7_ISSUER_AND_SERIAL, serial, ASN1_INTEGER)
+} ASN1_SEQUENCE_END(PKCS7_ISSUER_AND_SERIAL)
+
+IMPLEMENT_ASN1_FUNCTIONS(PKCS7_ISSUER_AND_SERIAL)
+
+/* Minor tweak to operation: free up X509 */
+static int ri_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
```

Comment:
Interesting callback mechanism, should we also take the chance to improve on OpenSSL's static function naming clearer for here and below?

Minor nit for documentation style for here and below:
```suggestion
// Minor tweak to operation: free up X509.
static int recip_info_cb(int operation, ASN1_VALUE **pval, const ASN1_ITEM *it,
```

Reply:
one goal i had in this was to preserve as much similarity as possible. between this series and OpenSSL's code. that said, this is just (static) function name and comment style, so yeah that works.